src/migrant/database.gleam

```diff
@@ -135,73 +161,141 @@ fn run_migration(
+  db: sqlight.Connection,
+) -> Result(Nil, Error) {
   let #(name, migration) = migration_tuple
-  case migration.up {
-    Some(sql) -> {
-      io.println("-> Applying migration: " <> name)
-      case exec(db, sql) {
-        Ok(_) -> {
-          case mark_migration_as_applied(db, migration_tuple) {
-            Ok(_) -> Ok(Nil)
-            Error(e) -> {
-              io.println("-> Failed to mark migration as applied: " <> name)
-              io.debug(e)
-              io.println("-> Rolling back migration: " <> name)
-
-              case rollback(migration_tuple, db) {
-                Ok(_) ->
-                  Error(MigrationError(
-                    "Rollback complete, failed to mark migration as applied: "
-                      <> name,
-                    e,
-                  ))
-                Error(e) -> {
-                  io.println("-> Failed to rollback migration: " <> name)
-                  Error(e)
-                }
-              }
-            }
-          }
-          Ok(Nil)
-        }
-        Error(e) -> {
-          io.println("-> Failed to apply migration: " <> name)
-          io.debug(e)
-          io.println("-> Rolling back migration: " <> name)
-
-          case rollback(migration_tuple, db) {
-            Ok(_) ->
-              Error(MigrationError(
-                "Rollback complete, failed to apply migration: " <> name,
-                e,
-              ))
-            Error(e) -> Error(e)
-          }
-        }
-      }
-    }
-    None -> {
-      io.println("-> Skipping migration: " <> name <> " no `up` query")
-      Ok(Nil)
-    }
+
+  use <- bool.guard(when: option.is_none(migration.up), return: {
+    io.println("-> Skipping migration: " <> name <> " no `up` query")
+    Ok(Nil)
+  })
+
+  io.println("-> Applying migration: " <> name)
+
+  use _ <- result.try(
+    db
+    |> exec("BEGIN TRANSACTION;")
+    |> with_err_message("Failed to begin transaction"),
+  )
+
+  // Attept to run all the actual UP queries that can fail
+  let res = {
+    // Execute the migration
+    use _ <- result.try(
+      exec(db, migration.up |> option.unwrap(""))
+      |> with_err_message("Failed to apply migration: " <> name),
+    )
+
+    // Attempt to mark the migration as applied
+    use _ <- result.try(
+      mark_migration_as_applied(db, migration_tuple)
+      |> with_err_message("Failed to mark migration as applied: " <> name),
+    )
+
+    // Attempt to commit the transaction
+    use _ <- result.try(
+      db
+      |> exec("COMMIT;")
+      |> with_err_message("Failed to commit transaction"),
+    )
+
+    Ok(Nil)
   }
+
+  // If the migration was successful, return
+  use <- bool.guard(when: result.is_ok(res), return: {
+    io.println("-> Migration applied successfully: " <> name)
+    Ok(Nil)
+  })
+
+  // If the migration failed and we don't have a down query, rollback the transaction
+  let assert Error(err) = res
+  use <- bool.guard(when: option.is_none(migration.down), return: {
```

Comment:
Same problem as above. This would probably crash the program if the user provides a `down` transaction

```suggestion
  use <- bool.lazy_guard(when: option.is_none(migration.down), return: fn () {
```